algorithms/rainbow/agent.py

```diff
@@ -0,0 +1,158 @@
+"""
+Adapted from https://github.com/Kaixhin/Rainbow
+
+Wrapper class including setup, training and evaluation functions for Rainbow DQN model
+"""
+import os
+import numpy as np
+import torch
+from torch import optim
+
+from algorithms.rainbow.model import RainbowDQN
+
+
+class Agent:
+    """
+    Wraps control between both online and target network for setup, training and evaluation
+    """
+    def __init__(self, args, env):
+        """
+        Q(s,a) is the expect reward. Z is the full distribution from which Q is generated
+        Support represents the support of Z distribution
+        Z is represented with a fixed number of "atoms", which are discrete positions equidistant
+        along its support defined between Vmin-Vmax.
+
+        As an example, for a given (s,a) pair, we can represent Z(s,a) with 8 atoms as follows:
+
+                   .        .     .
+                .  |     .  |  .  |
+                |  |  .  |  |  |  |  .
+                |  |  |  |  |  |  |  |
+           Vmin ----------------------- Vmax
+        """
+        self.action_space = env.action_space
+        self.atoms = args.atoms
+        self.Vmin = args.V_min
+        self.Vmax = args.V_max
+        self.support = torch.linspace(args.V_min, args.V_max, self.atoms).to(device=args.device)
+        self.delta_z = (args.V_max - args.V_min) / (self.atoms - 1)
+        self.batch_size = args.batch_size
+        self.n = args.multi_step
+        self.discount = args.discount
+
+        self.online_net = RainbowDQN(args, self.action_space).to(device=args.device)
+        if args.model and os.path.isfile(args.model):
+            """
+            Always load tensors onto CPU by default, will shift to GPU if necessary to avoid 
+            GPU RAM surge when loading a model checkpoint as recommended in pytorch official 
+            documentation.
+            Source: https://pytorch.org/docs/stable/torch.html#torch.load
+            """
+            self.online_net.load_state_dict(torch.load(args.model, map_location='cpu'))
+        self.online_net.train()
+
+        self.target_net = RainbowDQN(args, self.action_space).to(device=args.device)
+        self.update_target_net()
+        self.target_net.train()
+        for param in self.target_net.parameters():
+            param.requires_grad = False
+
+        self.optimiser = optim.Adam(self.online_net.parameters(), lr=args.lr, eps=args.adam_eps)
+
+    def reset_noise(self):
+        """ Resets noisy weights in all linear layers (of online net only) """
+        self.online_net.reset_noise()
+
+    def act(self, state):
+        """ Acts based on single state (no batch) """
+        with torch.no_grad():
+            return (self.online_net(state.unsqueeze(0)) * self.support).sum(2).argmax(1).item()
```

Comment:
can you go through this with me?
online net returns z or q? `q = F.softmax(q, dim=2)` is what it returns. what shape is this? put a comment in agent.py on what shape this is and say z if it is z.

ok now, why do we multiply by self.support? and why do we `.sum(2)`?

Reply:
We covered all this in the learn function of agent.py. Please read that again and we can talk if there is any point to clarify. Should I make a comment saying `for further information go to learn function in agent.py`? 
Z(s,a) is the distribution of expected returns, Q is the mean of Z or simply expected return for each action. Z has size s*a*atoms and Q has size s*a.

Reply:
hmm. I actually had to go to model.py to look further.
So online_net returns Q and nothing here involves Z then. Why are you mentioning Z to me?
And again, why do we multiply by self.support? and why do we `.sum(2)`?

Reply:
Ok let's go slower. Z(s,a) is a distribution defined by the atoms ((x_i, p_i), ...., (x_n, p_n)). Since Q is the expectation of Z, i.e. its "mean", by definition you sum the products x_i * p_i to obtain it. Since Z has shape (1, a, num_atoms), multiplying by the x_is (support) with p_is (values) and summing the second axis gives you the Q for every action in the input state. Argmax is the greedy part. Let me know if you still have problems with this.